### PR TITLE
Classes

### PR DESCRIPTION
**Pull request by dbakovich**
MixBuilder page is used to override standard 'view' button
EditMix page - 'edit'&'new' buttons

MixBuilderController, Mixes, Songs, SongsSelector, SongWrapper, Tracks are custom created classes

### CODE REVIEW

File: classes/MixBuilderController.cls [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            {
+                Decimal length = 0;
+                Set<SelectOption> filters = new Set<SelectOption>();
+                filters.add(new SelectOption('--Any Genre--', '--Any Genre--'));
+                List<Song__c> songList = new List<Song__c>();
+                if(mode == 'view')
+                {
+                    if (String.isNotEmpty(filter) && filter != '--Any Genre--')
+                    {
+                        songList = [Select Name, Genre__c,Length_m__c  FROM Song__c Where Id in (Select Song__c From Track__c Where Mix__c = :mix.Id) And Genre__c = :filter];
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: use style guide

File: classes/MixBuilderController.cls [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        get
+        {
+            if (setCon == null)
+            {
+                Decimal length = 0;
+                Set<SelectOption> filters = new Set<SelectOption>();
+                filters.add(new SelectOption('--Any Genre--', '--Any Genre--'));
+                List<Song__c> songList = new List<Song__c>();
+                if(mode == 'view')
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: such 'view' should be static variables

File: classes/MixBuilderController.cls [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        get; set;
+    }
+
+    public Map<Id, Song__c> songsInMix
+    {
+        get; set;
+    }
+
+    public List<SelectOption> getFilters
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: all these variables should be better placed before constructor

Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: also use lazy initialization when possible

File: classes/MixBuilderController.cls [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        for (Song__c song : selectedList)
+        {
+            if (!songsId.contains(song.Id))
+            {
+                Track__c track = new Track__c(Mix__c = this.Mix.Id, Song__c = song.Id);
+                tracksForInsertion.add(track);
+            }
+        }
+
+        if (tracksForInsertion.size() > 0) insert tracksForInsertion;
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: where is {} ?

File: classes/MixBuilderController.cls [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    public list<Song__c> selectedList {get;set;}
+
+    public Pagereference save()
+    {
+        getSelectedSong();
+        selectedList = selectedSongs.values();
+        List<Track__c> tracksForDeletion = new List<Track__c>();
+        List<Track__c> tracksForInsertion = new List<Track__c>();
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: "newTracks" would be better

File: classes/MixBuilderController.cls [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        }
+    }
+
+    public list<Song__c> selectedList {get;set;}
+
+    public Pagereference save()
+    {
+        getSelectedSong();
+        selectedList = selectedSongs.values();
+        List<Track__c> tracksForDeletion = new List<Track__c>();
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: "removedTracks"

File: classes/MixBuilderController.cls [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        Decimal length = 0;
+        this.mix = (Mix__c)stdController.getRecord();
+        Map<String,String> PageParameters = new Map<String, String>();
+        PageParameters = ApexPages.currentPage().getParameters();
+        if (PageParameters.get('id') == null) 
+        {
+            mode = 'new';
+        }
+        else if (PageParameters.get('id') != null & PageParameters.get('retURL') != null) 
+            {
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: check tabs

File: classes/MixBuilderController.cls [hunk 8]
```diff
@@ -0,0 +1,265 @@
+public class MixBuilderController {
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: sharing/without sharing

File: classes/MixBuilderController.cls [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    
+    public Mix__c mix
+    {
+        get; set;
+    }
+
+    public MixBuilderController(ApexPages.StandardController stdController)
+    {
+        Decimal length = 0;
+        this.mix = (Mix__c)stdController.getRecord();
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: what if record is new?

File: classes/MixBuilderController.cls [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+            {
+                wrapperList.add(new SongWrapper(song, false));
+            }
+        }
+        return wrapperList;
+    }
+
+    public void getSelectedSong()
+    {
+        if (wrapperList!=null)
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: style guide

File: classes/MixBuilderController.cls [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        {
+            for(SongWrapper sw : wrapperList)
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: please, use more informative variable variant instead of "sw"

File: classes/MixBuilderController.cls [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+                }
+                else
+                {
+                    selectedSongs.remove(sw.song.id);
+                }
+            }
+        }
+    }
+
+    public list<Song__c> selectedList {get;set;}
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: style guide

File: classes/MixBuilderController.cls [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+        List<Track__c> tracksForInsertion = new List<Track__c>();
+        Set<Id> songsId = new Set<Id>();
+        
+        for (Track__c track : [SELECT Song__c FROM Track__c Where Mix__c = :mix.Id])
+        {
+          if (!selectedSongs.containsKey(track.Song__c)) 
+          {
+              tracksForDeletion.add(track);
+          }
+          else songsId.add(track.Song__c);
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: {}

File: classes/MixBuilderController.cls [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        get; set;
+    }
+
+    public Integer noOfRecords
+    {
+        get; set;
+    }
+
+    public Integer size
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: size of what?

File: classes/MixBuilderController.cls [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    public List<Song__c> songs 
+    { 
+        get
+        {
+            return (List<Song__c>)setCon.getRecords();
+        } 
+        set;
+    }
+
+    public void applyFilter()
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: filter of what?

File: classes/MixBuilderController.cls [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        {
+            if (setCon == null)
+            {
+                Decimal length = 0;
+                Set<SelectOption> filters = new Set<SelectOption>();
+                filters.add(new SelectOption('--Any Genre--', '--Any Genre--'));
+                List<Song__c> songList = new List<Song__c>();
+                if(mode == 'view')
+                {
+                    if (String.isNotEmpty(filter) && filter != '--Any Genre--')
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: static variable

File: classes/MixBuilderController.cls [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+                       }
+                       mixLength = length;
+                    remainingLength = 100 - mixLength;
+                    remainingTracks = 20 - trackCount;
+                }
+                setCon = new ApexPages.StandardSetController(songList);
+                setCon.setPageSize(size);
+                noOfRecords = setCon.getResultSize();
+                
+                if (noOfRecords == null)
```
Comment (ann-petrova-itechart) [inline classes/MixBuilderController.cls]: integer -> null ?

File: classes/MixBuilderController.cls (hunk 4 above)
Comment (dbakovich) [reply classes/MixBuilderController.cls]: it is necessary to use brackets even if we implement only one operation?

File: classes/MixBuilderController.cls (hunk 17 above)
Comment (dbakovich) [reply classes/MixBuilderController.cls]: ma fault, you are right:)

**Pull request merged**